### PR TITLE
NAVAND-552: calculate the number of instructions to download

### PR DESCRIPTION
**Pull request by dzinad**
### Description
Started implementing pre-downloading voice instructions.
This PR introduces an algorithm that can calculate the number of instructions that should be downloaded based on current RouteProgress.

### CODE REVIEW

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/predownloading/TimeBasedNextVoiceInstructionsProvider.kt [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+import com.mapbox.api.directions.v5.models.VoiceInstructions
+import com.mapbox.navigation.base.trip.model.RouteProgress
+import com.mapbox.navigation.utils.internal.logW
+
+internal class TimeBasedNextVoiceInstructionsProvider(
+    private val observableTimeSeconds: Int,
+) : NextVoiceInstructionsProvider {
+
+    override fun getNextVoiceInstructions(
```
Comment (dzinad) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/predownloading/TimeBasedNextVoiceInstructionsProvider.kt:11]: I was thinking about proxying this work to a background thread but I measured the time required for these calculations and results are the following: for the route from Munich to Toledo with observableTime being equal to 1 hour the maximum time during which this method was executing is 3 ms. Considering that the route is quite long and the default observableTim will be much less (3 minutes), I think we're good to leave these calculations on the main thread.

File: libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProvider.kt [hunk 2]
```diff
@@ -0,0 +1,67 @@
+package com.mapbox.navigation.ui.voice.api
+
+import com.mapbox.api.directions.v5.models.VoiceInstructions
+import com.mapbox.navigation.utils.internal.logW
+
+internal class TimeBasedNextVoiceInstructionsProvider(
+    private val observableTimeSeconds: Int,
```
Comment (VysotskiVadim) [inline libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProvider.kt:7]: do I understand correctly that it means "from how much seconds ahead voice instructions should be downloaded"?

Comment (dzinad) [reply libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProvider.kt:7]: Yes. Your option is more clear but I'm afraid I'm gonna have some MaxLength issues with that name. :)

File: libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProvider.kt [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+            ?.getOrNull(progress.stepIndex)
+            ?: return emptyList<VoiceInstructions>().also {
+                logW(
+                    "Route does not contain valid step, progress=$progress",
+                    LOG_CATEGORY
+                )
+            }
+
+        val voiceInstructions = mutableListOf<VoiceInstructions>()
+        var cumulatedTime = progress.stepDurationRemaining
```
Comment (VysotskiVadim) [inline libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProvider.kt:28]: Maybe?
```suggestion
        var timeAheadOfProcessedVoiceInstructions = progress.stepDurationRemaining
```

File: libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProvider.kt [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+                    } else {
+                        currentStep = legSteps.first()
+                    }
+                } else {
+                    break
+                }
+            }
+            currentStep.voiceInstructions()?.let { voiceInstructions.addAll(it) }
+            cumulatedTime += currentStep.duration()
+        }
```
Comment (VysotskiVadim) [inline libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProvider.kt:60]: Maybe?
```suggestion
        findStepsAhead(observableTimeSeconds - cumulatedTime, progress).forEach {
            currentStep.voiceInstructions()?.let { voiceInstructions.addAll(it) }
        }
```
where
```
fun findStepsAhead(timeAhead: Int, progress: RouteProgressData): List<RouteStep> {
        val result = mutableListOf<RouteStep>()
        var currentStepIndex = progress.stepIndex
        var currentLegIndex = progress.legIndex
        while (cumulatedTime < observableTimeSeconds) {
            if (currentStepIndex + 1 < (legSteps?.size ?: 0)) {
                currentStep = legSteps!![currentStepIndex + 1]
                currentStepIndex++
            } else {
                currentStepIndex = 0
                if (currentLegIndex + 1 < legs.size) {
                    legSteps = legs[currentLegIndex + 1].steps()
                    currentLegIndex++
                    if (legSteps.isNullOrEmpty()) {
                        continue
                    } else {
                        currentStep = legSteps.first()
                    }
                } else {
                    break
                }
            }
            result.add(currentStep)
            cumulatedTime += currentStep.duration()
        }
}
```

File: libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProvider.kt [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        val voiceInstructions = mutableListOf<VoiceInstructions>()
+        var cumulatedTime = progress.stepDurationRemaining
+        val currentStepInstructions = currentStep.voiceInstructions()?.filter { instruction ->
+            val distanceAlongGeometry = instruction.distanceAlongGeometry()
+            distanceAlongGeometry != null
+                && distanceAlongGeometry <= progress.stepDistanceRemaining
+        }
+        if (currentStepInstructions != null) {
+            voiceInstructions.addAll(currentStepInstructions)
+        }
```
Comment (VysotskiVadim) [inline libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProvider.kt:36]: Maybe extract a function `getNotPassedVoiceInstuctions(step, progress.distanceRemaining)`?

File: libnavui-voice/src/test/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProviderTest.kt [hunk 6]
```diff
@@ -0,0 +1,752 @@
...
+import com.mapbox.navigation.testing.LoggingFrontendTestRule
+import io.mockk.every
+import io.mockk.mockk
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import org.junit.Assert.assertEquals
+import org.junit.Rule
+import org.junit.Test
+
+@OptIn(ExperimentalCoroutinesApi::class)
+class TimeBasedNextVoiceInstructionsProviderTest {
```
Comment (VysotskiVadim) [inline libnavui-voice/src/test/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProviderTest.kt:16]: What do you think about having more specific tests like this:
```
val route = loadRealRouteFromJson()
val instructionsToDownload = provider.getNextVoiceInstructions(/* provide some specific place on the route */)
assertEquals(listOf("instruction 1", "instruction 2"), instructionsToDownload.map { it.announcement }
```

this would demonstrate how this work on read data/cases. What do you think about it?

**Comment by VysotskiVadim** [comment]
@dzinad , that will be a great improvement 👍 
Left a few comments to discuss implementation details a bit 🙂

File: libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProvider.kt (hunk 3 above)
Comment (dzinad) [reply libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProvider.kt:28]: But it's exactly the time that has already been cumulated. It increases when I process new steps.

File: libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProvider.kt (hunk 4 above)
Comment (dzinad) [reply libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProvider.kt:60]: This code does not compile and I don't really see what you're suggesting.

File: libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProvider.kt (hunk 5 above)
Comment (dzinad) [reply libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProvider.kt:36]: `distanceRemaining` is not used for any steps other than current. For the next steps the logic is a bit different. So the method would not be reused. Why extract then? I think it would only complicate things.

File: libnavui-voice/src/test/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProviderTest.kt (hunk 6 above)
Comment (dzinad) [reply libnavui-voice/src/test/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProviderTest.kt:16]: This approach would not allow us to test corner cases. And there are a loooot of them. I've changed the code 5+ times when I was writing these tests and these changes were improvements. 
With the real data you only cover the very base scenarios. And also with these huge jsons it's very hard to fix tests when they fail later.
It's a unit test so I'm in favour of simple data with lots of corner cases. Sanity check with big data can be performed in either instrumentation tests or when running examples.

File: libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProvider.kt [hunk 7]
```diff
@@ -0,0 +1,67 @@
+package com.mapbox.navigation.ui.voice.api
+
+import com.mapbox.api.directions.v5.models.VoiceInstructions
+import com.mapbox.navigation.utils.internal.logW
+
+internal class TimeBasedNextVoiceInstructionsProvider(
+    private val observableTimeSeconds: Int,
+) : NextVoiceInstructionsProvider {
+
+    override fun getNextVoiceInstructions(progress: RouteProgressData): List<VoiceInstructions> {
```
[the remark this answers is not in the document]
Comment (dzinad) [reply libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProvider.kt:10]: The implementation details are in the doc references in the ticket.
TLDR: yes, it will be called on RouteProgress updates but not on every update. By default the interval will be `0.5 * observableTimeSeconds`.
The same instructions will not be downloaded twice but it's not for this class to handle.
I see your point about duplicated work here but I'm something can or should be done. For 2 reasons:
1) There's relatively little work here: it's not downloading, file reading or sth like that;
2) The instructions will be downloaded in parallel so we don't know if the latest downloaded instruction is the last one or not. Even if it is, what if some previous instructions failed to be downloaded? I think we'd want to retry the request then. So yeah, before making the request already downloaded stuff will definitely be filtered out but I don't think any changes are required here.

File: libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProvider.kt (hunk 4 above)
Comment (VysotskiVadim) [reply libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProvider.kt:60]: > This code does not compile

sorry for that.

I had difficulties understanding what's going on there. I understood it as following:
1. iteration over steps from different legs until observableTimeSeconds is reached  
2. add all voice instructions from the steps to result

So I decided to suggest extract the iteration logic into a function to simplify readability. Readability is a subjective topic, so all I'm doing is sharing opinion of how to make code more readable for me

File: libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProvider.kt (hunk 3 above)
Comment (VysotskiVadim) [reply libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProvider.kt:28]: I didn't understood what it does at the beginning so when I read and understood the code I decided to suggest a name which would help me understood the algorithm easier at the beginning. This is subjective topic, all I'm saying is that this would be more understandable for me. I'm okay if you don't agree and keep current name

File: libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProvider.kt (hunk 5 above)
Comment (VysotskiVadim) [reply libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProvider.kt:36]: > Why extract then? 

functions's names could be used as a documentation. Reading function's name, a programmer can understand what to expect in the code. It also improve understanding when you can observe algorithms from different levels. For example, here's how it can be represented on the hight level:
```kotlin
val result = findNotPassedVoiceInstructionsFromCurrentStep(...) + findStepsAheadInTime(observableTimeSeconds).map { it.voiceInstruction }
```
And then, if you want to check specific details, you can dive into the functions to see how exactly it finds instructions from the current step, or how it iterates over the steps

Comment (dzinad) [reply libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProvider.kt:36]: I'll try to think of something.

File: libnavui-voice/src/test/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProviderTest.kt (hunk 6 above)
Comment (VysotskiVadim) [reply libnavui-voice/src/test/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProviderTest.kt:14]: Just in case, I'm not trying to say that it's something wrong with existing tests, they are great! 👍

I think that one test with real data would be a great addition to exiting ones. It would prove that everything works on the real data as well. What's more, it could be used as an example of how the feature works: "if you're here on this route the feature downloads following voice instructions". It's similar to how I would test it manually on examples, but fast 🙂 

It's okay to test with real data in instrumentation tests, but there is nothing android specific, why not to test in on fast JVM? 🤔 

Anyway, it's not a strong opinion, just a suggestion. If you ignore it, I'm fine with it. Current tests suite is great already 👍

**Review by VysotskiVadim — APPROVED** [review]
The PR looks good enough for me 👍 
It would be nice to split the algorithm into a smaller functions to improve readability, but I'm okay if it will be done later. Let's carry on iterating 💪
